### PR TITLE
RK-12589 - Go AWS Lambda

### PR DESCRIPTION
**Pull request by GuyGazit**
Two examples for creating a Go AWS Lambda with Rookout:
* by uploading a zip with the binary
* by uploading a docker image

### CODE REVIEW

File: go-aws-lambda-docker/README.md [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+RUN go build -tags=alpine314,rookout_static -gcflags='all=-N -l' main.go
+
+FROM alpine:3.15 as release
+COPY --from=builder /app/main ./
+
+ENV ROOKOUT_TOKEN=${ROOKOUT_TOKEN}
+ENV ROOKOUT_CONTROLLER_PORT=443
+ENV ROOKOUT_CONTROLLER_HOST=wss://staging.control.rookout.com
+ENV ROOKOUT_DEBUG=1
```
Comment (ElDuderinos) [inline go-aws-lambda-docker/README.md:62]: We just need the token

File: go-aws-lambda-docker/README.md [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+ENV PORT 1994
+EXPOSE 1994
+CMD ["./main"]
+```
+
+The changes include:
+1. Two `go env` commands to replace the GoSDK with our GoRook.
```
Comment (ElDuderinos) [inline go-aws-lambda-docker/README.md:70]: Should be "...to properly use the Rookout artifactory"

File: go-aws-lambda-zip/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+ARG ARTIFACTORY_CREDS
+
+RUN go env -w GONOSUMDB="github.com/Rookout/GoSDK"
+RUN go env -w GOPROXY="https://proxy.golang.org,https://${ARTIFACTORY_CREDS}@rookout.jfrog.io/artifactory/api/go/rookout-go,direct"
+
+RUN apk --update --no-cache add git gcc musl-dev protobuf-dev openssl-libs-static openssl-dev build-base zlib-static
+
+# Change the version here to the newest one
+RUN go get -d github.com/Rookout/GoSDK@v0.1.10
```
Comment (ElDuderinos) [inline go-aws-lambda-zip/Dockerfile:14]: Align the version here with the one in the docker dockerfile example

File: go-aws-lambda-zip/README.md [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+## Rookout Integration Process
+
+### main.go
+Let's first change your main go file, there are 2 changes:
+1. Import the GoSDK like so: `rook "github.com/Rookout/GoSDK"`
+2. Add a call to our `rook.Start` at the start of your handler function: `rook.Start(rook.RookOptions{})` like so, and add a defer call to our `rook.Flush` like so: `defer rook.Flush()`
+You can look in `main.go` to see these lines in code.
+
+### Dockerfile
+Now we need to compile the code with the GoRook, for that we'll just use a docker.
```
Comment (ElDuderinos) [inline go-aws-lambda-zip/README.md:14]: Should be GoSDK, not GoRook

**Review by ElDuderinos — DISMISSED** [review]
Looks good overall, see some comments that I added.
Let's have also have @lirook review this? So we'll have a better end user point of view.

File: go-aws-lambda-docker/README.md [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+EXPOSE 1994
+CMD ["./main"]
+```
+
+The changes include:
+1. Two `go env` commands to properly use the Rookout artifactory.
+    * NOTE: Please contact us in order to get the artifactory credentials.
+2. The `apk` command to install tools required to build with the GoRook.
+3. `go get` to Install the GoRook.
+4. Adding `-tags=alpine314,rookout_static -gcflags='all=-N -l'` to the `go build`, this is needed for the GoRook to work.
```
Comment (lirook) [inline go-aws-lambda-docker/README.md:71]: Isn't `rookout_static` the default?

Comment (lirook) [reply go-aws-lambda-docker/README.md:71]: Also "GoRook"

File: go-aws-lambda-docker/README.md [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+5. ROOKOUT_TOKEN `ENV` is for convenience, you can add add manually to the Lambda function or pass them in the `RookOptions` in `main.go` instead.
```
Comment (lirook) [inline go-aws-lambda-docker/README.md:72]: You have a duplicate "add"?
Also, "pass them" - pass who
"In the RookOptions" -> "in the RookOptions variable"

File: go-aws-lambda-docker/README.md [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+
+### Uploading the Docker
+Now after adding those changes (and running `go mod init <name>` if you are creating a new go module) you need to upload the docker so you could use it as a lambda function.
```
Comment (lirook) [inline go-aws-lambda-docker/README.md:75]: Is creating a go module needed for Lambda? Why are we creating a go module?
docker -> Docker
lambda -> Lambda
what do you mean by "upload the docker"?

File: go-aws-lambda-docker/README.md [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+
+For this you need to run four commands:
+```bash
```
Comment (lirook) [inline go-aws-lambda-docker/README.md:78]: Maybe make the region a placeholder as well?

File: go-aws-lambda-docker/README.md [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+aws ecr get-login-password --region us-east-1 | docker login --username AWS --password-stdin <aws_account_id>.dkr.ecr.us-east-1.amazonaws.com
+docker build . --tag <docker_tag> --build-arg ARTIFACTORY_CREDS=<artifactory_credentials>
+docker tag <docker_tag>:latest <aws_account_id>.dkr.ecr.us-east-1.amazonaws.com/<docker_tag>:latest
+docker push <aws_account_id>.dkr.ecr.us-east-1.amazonaws.com/<docker_tag>
+```
+
+aws_account_id - Your account id on AWS.
+docker_tag - Just a tag to identify the docker you are building.
+artifactory_credentials - The credentials needed to get the GoRook from jfrog.
```
Comment (lirook) [inline go-aws-lambda-docker/README.md:87]: GoRook -> Rookout SDK
jfrog -> JFrog

File: go-aws-lambda-docker/README.md [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+For this you need to run four commands:
+```bash
+aws ecr get-login-password --region us-east-1 | docker login --username AWS --password-stdin <aws_account_id>.dkr.ecr.us-east-1.amazonaws.com
+docker build . --tag <docker_tag> --build-arg ARTIFACTORY_CREDS=<artifactory_credentials>
+docker tag <docker_tag>:latest <aws_account_id>.dkr.ecr.us-east-1.amazonaws.com/<docker_tag>:latest
+docker push <aws_account_id>.dkr.ecr.us-east-1.amazonaws.com/<docker_tag>
+```
+
+aws_account_id - Your account id on AWS.
+docker_tag - Just a tag to identify the docker you are building.
```
Comment (lirook) [inline go-aws-lambda-docker/README.md:86]: "the docker" - do you mean "the docker image"? 
Also docker -> Docker

File: go-aws-lambda-docker/main.go [hunk 11]
```diff
@@ -0,0 +1,35 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"time"
+
+	rook "github.com/Rookout/GoSDK"
```
Comment (lirook) [inline go-aws-lambda-docker/main.go:8]: can you change "rook" to "rookout"?

File: go-aws-lambda-zip/go.mod [hunk 12]
```diff
@@ -0,0 +1,10 @@
+module go-aws-lambda
+
+go 1.17
+
+require (
+	github.com/Rookout/GoSDK v0.1.10-stub
```
Comment (lirook) [inline go-aws-lambda-zip/go.mod:6]: Why do you require the stub?

**Review by lirook — CHANGES_REQUESTED** [review]
Looks good overall. I left comments and suggestions.
I would recommend search-and-replacing a few terms:
rook -> rookout
GoRook -> Rookout SDK
docker -> Docker
lambda -> Lambda

File: go-aws-lambda-docker/README.md [hunk 13]
```diff
@@ -0,0 +1,90 @@
+# Golang + AWS Lambda Docker Image Example
+
+A sample application for debugging Go + AWS Lambda using a docker image.
+
+## Rookout Integration Process
+
+Let's first change your main go file, there are 2 changes:
+1. Import the GoSDK like so: `rook "github.com/Rookout/GoSDK"`
+2. Add a call to our `rook.Start` at the start of your handler function: `rook.Start(rook.RookOptions{})` like so, and add a defer call to our `rook.Flush` like so: `defer rook.Flush()`
```
[the remark this answers is not in the document]
Comment (GuyGazit) [reply go-aws-lambda-docker/README.md:9]: The `defer` makes sure the `flush` runs at the end of the function, so no need

File: go-aws-lambda-docker/README.md (hunk 5 above)
Comment (GuyGazit) [reply go-aws-lambda-docker/README.md:71]: it is the default but we thought about keeping it incase the default changes

File: go-aws-lambda-docker/README.md (hunk 7 above)
Comment (GuyGazit) [reply go-aws-lambda-docker/README.md]: It is needed for go to be able to use different go modules, also made "upload the docker" more indicative

File: go-aws-lambda-zip/go.mod (hunk 12 above)
Comment (GuyGazit) [reply go-aws-lambda-zip/go.mod:6]: This is automatic stuff go does when you import the GoSDK module

Comment (lirook) [reply go-aws-lambda-zip/go.mod:6]: Not if you have the proxy set up and you "go get" a specific version.
If you leave it like this, it shouldn't work.

**Review by lirook — APPROVED** [review]

**Pull request merged**